_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,782 @@
+knitr::include_graphics("images/g_aleatoric_linear_8.png")
+```
+
+Using linear activations only, we also need to do more experimenting with the `scale = ...` line to get the result look "right". With `relu`, on the other hand, results are pretty robust to changes in how `scale` is computed. Which activation do we choose? If our goal is to adequately model variation in the data, we can just choose `relu` -- and leave assessing _uncertainty in the model_ to a different technique (the _epistemic_ uncertainty that is up next).
+
+Overall, it seems like aleatoric uncertainty is the straightforward part. We want the network to learn the variation inherent in the data, which it does. What do we gain? Instead of obtaining just point estimates, which in this example might turn out pretty bad in the two fan-like areas of the data on the left and right sides, we learn about the spread as well. We'll thus be appropriately cautious depending on what input range we're making predictions for.
+
+### Epistemic uncertainty
+
+Now our focus is on the model. Given a speficic model (e.g., one from the linear family), what kind of data does it say conforms to its expectations?
+
+To answer this question, we make use of a [variational-dense layer](https://rstudio.github.io/tfprobability/reference/layer_dense_variational.html).
+This is again a Keras layer provided by `tfprobability`. Internally, it works by minimizing the _evidence lower bound_ (ELBO),  thus striving to find an approximative posterior that does two things:
+
+(1) fit the actual data well (put differently: achieve high _log likelihood_), and
+(2) stay close to a _prior_ (as measured by [KL divergence](https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence)).
+
+As users, we actually specify the form of the posterior as well as that of the prior. Here is how a prior could look.
+
+```{r}
+prior_trainable <-
+  function(kernel_size,
+           bias_size = 0,
+           dtype = NULL) {
+    n <- kernel_size + bias_size
+    keras_model_sequential() %>%
+      # we'll comment on this soon
+      # layer_variable(n, dtype = dtype, trainable = FALSE) %>%
+      layer_variable(n, dtype = dtype, trainable = TRUE) %>%
+      layer_distribution_lambda(function(t) {
+        tfd_independent(tfd_normal(loc = t, scale = 1),
+                        reinterpreted_batch_ndims = 1)
+      })
+  }
+
+```
+
+
+This prior is itself a Keras model, containing a [ layer that wraps a variable](https://rstudio.github.io/tfprobability/reference/layer_variable.html) and a `layer_distribution_lambda`, that type of distribution-yielding layer we've just encountered above. The variable layer could be fixed (non-trainable) or non-trainable, corresponding to a genuine prior or a prior learnt from the data in an _empirical Bayes_-like way. The distribution layer outputs a normal distribution since we're in a regression setting.
```

Comment:
> fixed (non-trainable) or non-trainable

This is a little confusing to me. Are the two options: (1) fixed non-trainable, (2) non-trainable?